services/device-registry-mongodb/src/main/java/org/eclipse/hono/deviceregistry/mongodb/service/MongoDbBasedTenantService.java

```diff
@@ -0,0 +1,514 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.stereotype.Component;
+
+import io.opentracing.Span;
+import io.opentracing.noop.NoopSpan;
+import io.vertx.core.AbstractVerticle;
+import io.vertx.core.Future;
+import io.vertx.core.Promise;
+import io.vertx.core.json.JsonObject;
+import io.vertx.ext.mongo.FindOptions;
+import io.vertx.ext.mongo.IndexOptions;
+import io.vertx.ext.mongo.MongoClient;
+import io.vertx.ext.mongo.UpdateOptions;
+
+/**
+ * An implementation of the tenant service and the tenant management service
+ * that uses a mongodb database to store tenants.
+ * <p>
+ * On startup this adapter tries to find the tenant collection, if not found, it gets created.
+ *
+ * @see <a href="https://www.eclipse.org/hono/docs/api/tenant/">Tenant API</a>
+ * @see <a href="https://www.eclipse.org/hono/docs/api/management/">Device Registry Management API</a>
+ */
+@Component
+@Qualifier("serviceImpl")
+public final class MongoDbBasedTenantService extends AbstractVerticle
+        implements TenantService, TenantManagementService {
+
+    private static final Logger log = LoggerFactory.getLogger(MongoDbBasedTenantService.class);
+
+    private static final int INDEX_CREATION_MAX_RETRIES = 3;
+    private MongoClient mongoClient;
+    private MongoDbCallExecutor mongoDbCallExecutor;
+    private MongoDbBasedTenantsConfigProperties config;
+
+    /**
+     * Sets an instance of the {@code MongoDbCallExecutor} and {@code MongoClient}.
+     *
+     * @param mongoDbCallExecutor An instance of the mongoDbCallExecutor.
+     * @throws NullPointerException if the mongoDbCallExecutor is {@code null}.
+     */
+    @Autowired
+    public void setExecutor(final MongoDbCallExecutor mongoDbCallExecutor) {
+        this.mongoDbCallExecutor = Objects.requireNonNull(mongoDbCallExecutor);
+        this.mongoClient = Objects.requireNonNull(this.mongoDbCallExecutor.getMongoClient());
+    }
+
+    /**
+     * Sets the configuration properties for this service.
+     *
+     * @param config The configuration properties.
+     * @throws NullPointerException if the configuration is {@code null}.
+     */
+    @Autowired
+    public void setConfig(final MongoDbBasedTenantsConfigProperties config) {
+        this.config = Objects.requireNonNull(config);
+    }
+
+    @Override
+    public void start(final Promise<Void> startPromise) {
+        mongoDbCallExecutor.createCollectionIndex(config.getCollectionName(),
+                new JsonObject().put(RegistrationConstants.FIELD_PAYLOAD_TENANT_ID,
+                        1),
+                new IndexOptions().unique(true), INDEX_CREATION_MAX_RETRIES)
+                .map(success -> {
+                    startPromise.complete();
+                    return null;
+                })
+                .onFailure(startPromise::fail);
```

Comment:
One more index for the _subjectDN_ is optimal, so that it helps with the querying based on the _subjectDNs_. Please refer my comment on the update Tenant part and the same index should work.

Reply:
done